spanconfig: modify existing RPCs to work with system span configurations

arulajmani opened the pull request:
See individual commits for details.

arulajmani committed:
roachpb: modify SpanConfigEntry and span config RPCs to work on targets

This patch modifies roachpb.SpanConfigEntry to tie a
roachpb.SpanConfigTarget (instead of roachpb.Span)  with a
roachpb.SpanConfig.

A roachpb.SpanConfigTarget is a union proto over roachpb.Span and
roachpb.SystemSpanConfigTarget. Given a roachpb.SpanConfigEntry,
we can use the roachpb.SpanConfigTarget to disambiguate between
regular span configurations and system span configurations.

We use the modified roachpb.SpanConfigEntry in all our RPCs. We also
replace usages of roachpb.Spans with roachpb.SpanConfigTargets in the
RPCs as well.

Release note: None

arulajmani committed:
spanconfig: introduce SystemTarget and encoding/decoding methods

This patch introduces a new spanconfig.SystemTarget type. When used
in conjunction with a span configuration, tenants (and the system
tenant) can use this type to target multiple keyspans with the
 associated span configuration. We use this type to expand the
a spanconfig.Target to optionally embedd a SystemTarget (in additon
to a span target like it did before).

We also add special encoding/decoding methods for system targets, which
we will make use of when persisting them in
`system.span_configurations`. We do so by reserving a special
SystemSpanConfigSpan at the end of the System range. This allows us to
then assign special meaning to key prefixes carved out of this range
and the corresponding spans stored in `system.span_configurations`. We
must take special care that the host tenant doesn't install span
configurations to these keyspans directly, and as such, we make
exceptions in the translator and reconciler.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76219)
<!-- Reviewable:end -->

arulajmani commented:
Rebased on top of master now that #76213 has landed. 

@nvanbenschoten @ajwerner, this was the PR I called out in yesterday's pod meeting -- the second commit, specifically around the encoding/decoding logic + the key span we're reserving for system span configurations was something I was hoping to get your eyes on.

adityamaru commented on pkg/roachpb/span_config.proto:
```diff
@@ -166,56 +166,99 @@ message SpanConfig {
+	TenantID SourceTenantID = 1 [(gogoproto.nullable) = false];
+
+	// TargetTenantID is the ID of the tenant that the associated system span
+	// configuration applies to. If the host tenant is the source then the
```
nit: Can we reword the sentence "If the host tenant is the source then the system span config applies over all ranges in the system (including those belonging to secondary tenants)." to mention that both SourceTenantID and TargetTenantID have to be set to the host tenant ID.

adityamaru commented on pkg/roachpb/span_config.proto:
```diff
@@ -166,56 +166,99 @@ message SpanConfig {
+
+  // Targets to request configurations for. The targets listed here are not
+  // allowed to be duplicated/overlap with one another.
+  repeated SpanConfigTarget targets = 3 [(gogoproto.nullable) = false];
```
nit: this should be field with ID 2?

adityamaru commented on pkg/roachpb/span_config.proto:
```diff
@@ -166,56 +166,99 @@ message SpanConfig {
+  //
+  // Any system span configurations set by the tenant are also returned if
+  // requested.
+  repeated SpanConfigEntry span_config_entries = 2 [(gogoproto.nullable) = false];
```
just for my understanding, why do we need to bump the ID on this field? The wire type is still the same right?

adityamaru commented on pkg/roachpb/span_config.proto:
```diff
@@ -166,56 +166,99 @@ message SpanConfig {
-  // ToUpsert captures the spans we want to upsert and the configs we want to
-  // upsert with.
-  repeated SpanConfigEntry to_upsert = 2 [(gogoproto.nullable) = false];
+  // To upsert captures the targets we want to upsert and the configs we want
```
nit: s/To upsert/ToUpsert/g

adityamaru commented on pkg/roachpb/span_config.proto:
```diff
@@ -166,56 +166,99 @@ message SpanConfig {
-  repeated SpanConfigEntry to_upsert = 2 [(gogoproto.nullable) = false];
+  // To upsert captures the targets we want to upsert and the configs we want
+  // to upsert with.
+  repeated SpanConfigEntry to_upsert = 4 [(gogoproto.nullable) = false];
```
ditto question as above, do we need to change this ID?

adityamaru commented on pkg/spanconfig/target.go:
```diff
@@ -15,29 +15,47 @@ import "github.com/cockroachdb/cockroach/pkg/roachpb"
+	span roachpb.Span
+}
+
+func MakeTarget(t roachpb.SpanConfigTarget) Target {
```
nit: lint will probably need a comment

adityamaru commented on pkg/spanconfig/target.go:
```diff
@@ -53,12 +71,27 @@ func (t Target) String() string {
 // encoding. It is the inverse of Encode.
 func DecodeTarget(sp roachpb.Span) Target {
-	return Target(sp)
+	return Target{
```
nit: i'd collapse this to one line, here and throughout but its a purely aesthetic comment so feel free to ignore.

adityamaru commented on pkg/rpc/auth_tenant.go:
```diff
@@ -299,18 +331,14 @@ func (a tenantAuthorizer) authUpdateSpanConfigs(
+	case *roachpb.SpanConfigTarget_SystemSpanConfigTarget:
+		return validateSystemTarget(*spanConfigTarget.GetSystemSpanConfigTarget())
+	default:
+		panic("unknown span config target type")
```
should we return an error instead of panicking?

adityamaru reviewed (commented):
Leaving some comments on the first commit, generally LGTM. I'm going to come back to the second commit in a little bit.

ajwerner reviewed (commented):
Didn't make it all the way through, publishing what I had

arulajmani reviewed (commented):
Addressed all the comments so far, RFAL! 

> Dismissed @adityamaru and @ajwerner from 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-:-MvZjXnu8JP2Qp0BFPPc:b9iqyk1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @irfansharif, and @nvanbenschoten)

___
*[pkg/roachpb/span_config.proto, line 178 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvMxNf6dmwJguHlaDTO-r1-178:-MvZHSQ631Ll-Eel-KSD:baf33ic) ([raw file](https://github.com/cockroachdb/cockroach/blob/36415d73f730b0862b62aaf126df6fca61d08aa1/pkg/roachpb/span_config.proto#L178)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

nit: Can we reword the sentence "If the host tenant is the source then the system span config applies over all ranges in the system (including those belonging to secondary tenants)." to mention that both SourceTenantID and TargetTenantID have to be set to the host tenant ID. 
</blockquote></details>

Good call, done. 
___
*[pkg/roachpb/span_config.proto, line 216 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvMxNf6dmwJguHlaDTO-r1-216:-MvZHd042lCbArTb5JE_:bwxumbr) ([raw file](https://github.com/cockroachdb/cockroach/blob/36415d73f730b0862b62aaf126df6fca61d08aa1/pkg/roachpb/span_config.proto#L216)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

nit: this should be field with ID 2?
</blockquote></details>

Yup, I had some other things here before but now they're gone. 
___
*[pkg/roachpb/span_config.proto, line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvMxNf6dmwJguHlaDTO-r1-230:-MvZHsTM9ZZRRaaIAktS:b-4c50hd) ([raw file](https://github.com/cockroachdb/cockroach/blob/36415d73f730b0862b62aaf126df6fca61d08aa1/pkg/roachpb/span_config.proto#L230)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

just for my understanding, why do we need to bump the ID on this field? The wire type is still the same right?
</blockquote></details>

The type is still the same, but the underlying type (`SpanConfigEntry`) changed. I think I'm being overly cautious in bumping things here (and in some other places with the bumping), given the already existing migration in conjunction with these being RPC types that aren't persisted. Do sanity check my claim here though. 
___
*[pkg/roachpb/span_config.proto, line 261 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvMxNf6dmwJguHlaDTO-r1-261:-MvZJIEV5iZ7scwIX5zI:b-txtf5v) ([raw file](https://github.com/cockroachdb/cockroach/blob/36415d73f730b0862b62aaf126df6fca61d08aa1/pkg/roachpb/span_config.proto#L261)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

ditto question as above, do we need to change this ID?
</blockquote></details>

See above. 
___
*[pkg/rpc/auth_tenant.go, line 340 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvMxNf6dmwJguHlaDTP-r1-340:-MvZK1mY13ysU_0AA2MT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/36415d73f730b0862b62aaf126df6fca61d08aa1/pkg/rpc/auth_tenant.go#L340)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

should we return an error instead of panicking?
</blockquote></details>

Done.
___
*[pkg/spanconfig/systemtarget.go, line 82 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvMxNfAcPPGVpHMv746-r2-82:-MvZPOiP1vnmA_eqSHHP:b5jfawc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b089a73d085076c9fc1b55d634e0cce750dab22/pkg/spanconfig/systemtarget.go#L82)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Should we be defensive about validating `EndKey`?
</blockquote></details>

Good call, done. Added a test that tries to decode invalid spans as well. 
___
*[pkg/spanconfig/systemtarget.go, line 85 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvMxNfAcPPGVpHMv746-r2-85:-MvZPKZ91UbHu68sJ9rP:b-lx4fj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b089a73d085076c9fc1b55d634e0cce750dab22/pkg/spanconfig/systemtarget.go#L85)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

should this be `HasPrefix` or should it be an equality check?
</blockquote></details>

Fixed. 
___
*[pkg/spanconfig/target.go, line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvMxNf8jP4NW9IFNqB2-r1-92:-MvZLLia4uY8F-2A1ZmI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/36415d73f730b0862b62aaf126df6fca61d08aa1/pkg/spanconfig/target.go#L92)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

nit: i'd collapse this to one line, here and throughout but its a purely aesthetic comment so feel free to ignore.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

adityamaru reviewed (approved):
Nice! Second commit also LGTM, the panics everywhere make me a little nervous so once we get rid of the ones which won't actually be panicking once this is built out further, we should either plumb proper error handling or ReportOrPanic instead?

arulajmani commented:
> Nice! Second commit also LGTM, the panics everywhere make me a little nervous so once we get rid of the ones which won't actually be panicking once this is built out further, we should either plumb proper error handling or ReportOrPanic instead?

I did some of error plumbing in the next PR https://github.com/cockroachdb/cockroach/pull/76414 for the case where we do need to. I'll just move them to the second commit here.

arulajmani reviewed (commented):
@adityamaru added a third commit on here to clean up the panics stuff if you want to give that a quick look as well. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-:-MvazXiVB4j4ZjDm8mUA:b-se6tn2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @irfansharif, and @nvanbenschoten)

___
*[pkg/spanconfig/target_test.go, line 144 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvV__yfHh_S-BTTELrU-r4-144:-Mvat-jD9YdrTQSTx842:b-4fu0me) ([raw file](https://github.com/cockroachdb/cockroach/blob/79867f82a82b360669fbdf3c52d0fe9a2cccddfe/pkg/spanconfig/target_test.go#L144)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

what is numOps for? Is it just to get 20 shuffles?
</blockquote></details>

Yup, just runs the thing 20 times. This is slightly overkill, but we'll rely on the sort order in the KVAccessor, so whatever :P 


<!-- Sent from Reviewable.io -->

adityamaru reviewed (approved):
Third commit LGTM, thanks for doing that.

nvb reviewed (commented):
Mostly nits. Nice job on this.

> Reviewed 3 of 9 files at r1, 39 of 39 files at r3, 31 of 34 files at r4, 3 of 3 files at r5, 7 of 7 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-:-MvpFTKm3elPw-vDsBzD:bw0bita)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @arulajmani, and @irfansharif)

___
*[pkg/roachpb/span_config.proto, line 173 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-Mvp9Hxp31b-aRuaX3G9:-Mvp9Hxp31b-aRuaX3GA:bhoxbzv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebec94f69bbf7e1b62f7d536606b737e9bebf081/pkg/roachpb/span_config.proto#L173)):*
> ```protobuf
>   option (gogoproto.equal) = true;
> 
> 	// SourceTenantID is the ID of the tenant that specified the system span
> ```

nit: is the spacing off here? Reviewable doesn't seem to like it.
___
*[pkg/roachpb/span_config.proto, line 175 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-Mvp9TzrEYUs8-ZUGM5X:-Mvp9TzrEYUs8-ZUGM5Y:basw620) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebec94f69bbf7e1b62f7d536606b737e9bebf081/pkg/roachpb/span_config.proto#L175)):*
> ```protobuf
> 	// SourceTenantID is the ID of the tenant that specified the system span
> 	// configuration.
> 	TenantID SourceTenantID = 1 [(gogoproto.nullable) = false];
> ```

Proto field names should be snake case, with an optional `gogoproto.customname`.
___
*[pkg/roachpb/span_config.proto, line 180 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvpA1l5Cadd85_DNeIR:-MvpA1l5Cadd85_DNeIS:bt1txps) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebec94f69bbf7e1b62f7d536606b737e9bebf081/pkg/roachpb/span_config.proto#L180)):*
> ```protobuf
> 	// configuration applies to.
> 	//
> 	// If the host tenant is both the source and the target then the system span
> ```

Does this mean that the system tenant can't target only itself? Is that desirable? Even if it's not undesirable, is it an intuitive special case?

Could we instead say that if the field is empty, the configuration applies to all tenants?
___
*[pkg/rpc/auth_tenant.go, line 314 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvpBLH-5kZQ2GmS6YQr:-MvpBLH-5kZQ2GmS6YQs:baxt5y1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebec94f69bbf7e1b62f7d536606b737e9bebf081/pkg/rpc/auth_tenant.go#L314)):*
> ```Go
> 		}
> 
> 		if target.SourceTenantID != tenID || target.SourceTenantID != target.TargetTenantID {
> ```

nit: Should these conditions be separated and return different errors? `SourceTenantID != tenID` means something different from `TargetTenantID != tenID`.
___
*[pkg/spanconfig/target.go, line 89 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvpFGzP0k0IggLYtBAp:-MvpFGzP0k0IggLYtBAq:bj9o1m8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/spanconfig/target.go#L89)):*
> ```Go
> 		log.Fatalf(ctx, "unknown type of system target %v", t)
> 	}
> 	return roachpb.Span{}
> ```

nit: you can change this to `panic("unreachable")` to avoid the confusing return value.
___
*[pkg/spanconfig/spanconfigtestutils/recorder.go, line 101 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvpCT13A3y9fBJsF5HD:-MvpCT13A3y9fBJsF5HE:bojmzoj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebec94f69bbf7e1b62f7d536606b737e9bebf081/pkg/spanconfig/spanconfigtestutils/recorder.go#L101)):*
> ```Go
> 			return mi.batchIdx < mj.batchIdx
> 		}
> 		if !mi.update.Target.Equal(mj.update.Target) { // sort by target ordere.
> ```

This comment looks off. `// sort by target order`


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, though not the deepest review I've ever given. 

> Reviewed 1 of 9 files at r1, 19 of 34 files at r4, 1 of 3 files at r5, 3 of 7 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-:-Mvspndd2OPJQBK0zVfg:b-geixy)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @arulajmani, and @irfansharif)

___
*[pkg/spanconfig/target.go, line 125 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvslKf50yKR_mPrdxln:-MvslKf50yKR_mPrdxlo:b-xaf4wb) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/spanconfig/target.go#L125)):*
> ```Go
> 	}
> 
> 	// We're dealing with 2 span targets; compare their start keys.
> ```

In the case that two spans have the same key, would we want to provide a total ordering over them based on the `EndKey`?



_Code quote:_
```Go
	// We're dealing with 2 span targets; compare their start keys.
	return t.GetSpan().Key.Compare(o.GetSpan().Key) < 0
```


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
TFTRs! 

Will merge on green. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-:-Mvtpo594wGyRf9xUPV9:bp2x2zx)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @ajwerner, @irfansharif, and @nvanbenschoten)

___
*[pkg/roachpb/span_config.proto, line 180 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvpA1l5Cadd85_DNeIR:-MvtR4zFAwwYlHxa_zQU:b-sgbili) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebec94f69bbf7e1b62f7d536606b737e9bebf081/pkg/roachpb/span_config.proto#L180)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this mean that the system tenant can't target only itself? Is that desirable? Even if it's not undesirable, is it an intuitive special case?

Could we instead say that if the field is empty, the configuration applies to all tenants?
</blockquote></details>

The special casing here was intended to mirror full cluster backup, which when run by the host tenant, backs-up all ranges (including tenant ranges). That being said, I do agree that this isn't an intuitive special case, especially at the level of these RPCs. 

Changed to make this field nullable + added auth validation to ensure only the host tenant is allowed to leave this field unset. 
___
*[pkg/rpc/auth_tenant.go, line 314 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvpBLH-5kZQ2GmS6YQr:-MvtUlT85Im7AQ4G9v2O:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebec94f69bbf7e1b62f7d536606b737e9bebf081/pkg/rpc/auth_tenant.go#L314)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: Should these conditions be separated and return different errors? `SourceTenantID != tenID` means something different from `TargetTenantID != tenID`.
</blockquote></details>

Done. 


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
> Reviewed 2 of 34 files at r4, 1 of 7 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76219)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @ajwerner, @arulajmani, @irfansharif, and @nvanbenschoten)

___
*[pkg/ccl/kvccl/kvtenantccl/connector.go, line 465 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvtiUQEBax7aamPJH2f:-MvtiUQF7lWJ4izI3DIc:b-1kx0zw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/ccl/kvccl/kvtenantccl/connector.go#L465)):*
> ```Go
> 	if err := c.withClient(ctx, func(ctx context.Context, c *client) error {
> 		resp, err := c.GetSpanConfigs(ctx, &roachpb.GetSpanConfigsRequest{
> 			Targets: spanconfig.TargetsToTargetProtos(ctx, targets),
> ```

Do we need to plumb down a context for a type-conversion
___
*[pkg/spanconfig/target.go, line 36 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvtjTOzBP4KdCL210yW:-MvtjTOzBP4KdCL210yX:bkf9gcx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/spanconfig/target.go#L36)):*
> ```Go
> 		// taught and tested the KVAccessor to work with system targets.
> 	default:
> 		return Target{}, errors.AssertionFailedf("unknown type of system target %v", t)
> ```

Getting rid of this error and opting instead for a panic would help simplify code paths (no handling of unhandle-able errors needed) that bottom out at what is affectively a type cast.
___
*[pkg/spanconfig/target.go, line 89 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvpFGzP0k0IggLYtBAp:-MvtjHIfAdnbxiI3peZ8:b5e69l4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/spanconfig/target.go#L89)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: you can change this to `panic("unreachable")` to avoid the confusing return value.
</blockquote></details>

+1. Would also remove the need to pass in a ctx for what should otherwise be a simple type cast.
___
*[pkg/spanconfig/target.go, line 102 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvtkL6U2wEZ9WfnEoKs:-MvtkL6VCXLNsl6eYP5_:bmroykw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/spanconfig/target.go#L102)):*
> ```Go
> 	// ID.
> 	if t.IsSystemTarget() && o.IsSystemTarget() {
> 		if t.GetSystemTarget().SourceTenantID == roachpb.SystemTenantID &&
> ```

Would this be simpler if we defined a .Less on the SystemTarget type itself? Ditto below for the equality check, where we delegate to span.Equal cause it's available. It seems generally useful to define these on the base type instead of inlining it in methods on the union type. That would for e.g. let us define sorts on the base types instead of roundtripping through the union type.
___
*[pkg/spanconfig/target.go, line 155 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvtkvqT98JsPz-73Xaf:-MvtkvqT98JsPz-73Xag:b-ewindu) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/spanconfig/target.go#L155)):*
> ```Go
> // isEmpty returns true if the receiver is an empty target.
> func (t Target) isEmpty() bool {
> 	if t.systemTarget.isEmpty() && t.span.Equal(roachpb.Span{}) {
> ```

[nit] This could be simplified to "return <condition>".
___
*[pkg/spanconfig/target.go, line 163 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-Mvtly2h0f9luaUipcUw:-Mvtly2h0f9luaUipcUx:b-4bgzjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/spanconfig/target.go#L163)):*
> ```Go
> // SpanConfigTargetProto returns a roachpb.SpanConfigTarget equivalent to the
> // receiver.
> func (t Target) SpanConfigTargetProto(ctx context.Context) roachpb.SpanConfigTarget {
> ```

Is it fair to say every spanconfig.Target is going to be represented as a roachpb.SpanConfigTarget over the wire? Ditto for []spanconfig.Targets represented as []roachpb.SpanConfigTarget? If so, [nit] how about the following method signatures?

```go
func (t Target) ToProto() roachpb.SpanConfigTarget { .. } // also dropping ctx and panic-ing instead
func TargetsToProtos([]Target) []roachpb.SpanConfigTarget { ...}
func TargetsFromProtos([]roachpb.SpanConfigTarget) []Target { ...}
```
___
*[pkg/spanconfig/target.go, line 216 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-Mvtl5ml1mKPNMB3g9cb:-Mvtl5ml1mKPNMB3g9cc:b-jz8fyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/spanconfig/target.go#L216)):*
> ```Go
> // RecordsToSpanConfigEntries converts a list of records to a list
> // roachpb.SpanConfigEntry protos suitable for sending over the wire.
> func RecordsToSpanConfigEntries(ctx context.Context, records []Record) []roachpb.SpanConfigEntry {
> ```

[nit] s/RecordsToSpanConfigEntries/RecordsToEntries for symmetry with EntriesToRecords below. Also nice to be a little less wordy.
___
*[pkg/spanconfig/spanconfigreconciler/reconciler.go, line 287 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvtnUMMARvhk6hOjo8r:-MvtnUMNC5fiIaufvJQB:b-mpt4ff) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/spanconfig/spanconfigreconciler/reconciler.go#L287)):*
> ```Go
> 		// reconciler we shouldn't be requesting these configs directly, instead,
> 		// they should be targeted through SystemSpanConfigTargets instead.
> 		targets = append(targets, spanconfig.MakeTargetFromSpan(roachpb.Span{
> ```

Does this smell to anyone? The server's internal encoding is leaking through to the client if it's having to be careful about not asking for portions of the keyspace. Ideally the server side code just filters these spans out at some level if we only want it to be targeted through explicit SystemSpanConfigTargets. Is this something we're planning to do anything about?
___
*[pkg/rpc/auth_test.go, line 153 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-Mvtp-Ry7GzW3h-xBgwd:-Mvtp-Ry7GzW3h-xBgwe:b-jv30hj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/rpc/auth_test.go#L153)):*
> ```Go
> 		return string(append(tenPrefix, []byte(key)...))
> 	}
> 	makeSpan := func(key string, endKey ...string) *roachpb.Span {
> ```

Can we avoid "pointer to a span" here as well? Ditto below for makeSystemTarget. For no other reasons than it being sharper to use, for unclear benefits. If we're doing union types, staring at makeSpanConfigTarget, we could instead have two constructors for each type of base type in the union. IIUC, one or the other argument _has_ to be nil, right? Could even repurpose makeSystemTarget and makeSpan for it.
___
*[pkg/rpc/auth_test.go, line 653 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-Mvtq1io5p9DQIyJwT42:-Mvtq1io5p9DQIyJwT43:b-5pnhsl) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/rpc/auth_test.go#L653)):*
> ```Go
> 			},
> 			{
> 				// Ensure tenant 10 (the tenant we test all these with) can't pretend
> ```

Is this the same test case as the last one? In fact, I'm seeing multiple duplicated sub-tests here. Was this intentional?


<!-- Sent from Reviewable.io -->

irfansharif reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-:-MvtvyJP4V-ulv7t84SL:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @ajwerner, @arulajmani, @irfansharif, and @nvanbenschoten)

___
*[pkg/roachpb/span_config.proto, line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvMxNf6dmwJguHlaDTO-r1-230:-MvttuB28SYV8XSvtzUM:bktibvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/36415d73f730b0862b62aaf126df6fca61d08aa1/pkg/roachpb/span_config.proto#L230)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

The type is still the same, but the underlying type (`SpanConfigEntry`) changed. I think I'm being overly cautious in bumping things here (and in some other places with the bumping), given the already existing migration in conjunction with these being RPC types that aren't persisted. Do sanity check my claim here though. 
</blockquote></details>

Bumping this isn't necessary, this is safe to revert (makes for a more confusing git commit history if I'm being pedantic).
___
*[pkg/roachpb/span_config.proto, line 180 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvpA1l5Cadd85_DNeIR:-MvtsatO6QLG65MNF4Y4:b-f9evdm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebec94f69bbf7e1b62f7d536606b737e9bebf081/pkg/roachpb/span_config.proto#L180)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

The special casing here was intended to mirror full cluster backup, which when run by the host tenant, backs-up all ranges (including tenant ranges). That being said, I do agree that this isn't an intuitive special case, especially at the level of these RPCs. 

Changed to make this field nullable + added auth validation to ensure only the host tenant is allowed to leave this field unset. 
</blockquote></details>

I think the following semantics are sound/easier to get wrong:
- Only the host tenant is allowed to set a target_tenant_id.
  - If the target_tenant_id is unset and the source_tenant_id == host, applies to all tenants.
  - If a target_tenant_id is set, it applies to the that tenant only (could be the host tenant itself)

This is in contrast to making it _necessary_ for secondary tenants to target themselves. In the alternative above, the whole concept of "targeting" a tenant doesn't extend to secondary tenants, which is what we want.
___
*[pkg/roachpb/span_config.proto, line 175 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvtsCaQ9ingHUR077oL:-MvtsCaQ9ingHUR077oM:b1oy27e) ([raw file](https://github.com/cockroachdb/cockroach/blob/a713d876ac558d6be4fd1a485e107af4f9b2f3f4/pkg/roachpb/span_config.proto#L175)):*
> ```protobuf
>   // SourceTenantID is the ID of the tenant that specified the system span
>   // configuration.
>   TenantID source_tenant_id = 1 [(gogoproto.nullable) = false, (gogoproto.customname)= "SourceTenantID"];
> ```

[nit] Spacing around '=' for gogoproto.customname. Ditto below, where the spacing is inconsistent in another way.
___
*[pkg/roachpb/span_config.proto, line 182 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvtsMXO4giQcTAFp4dd:-MvtsMXO4giQcTAFp4de:b-37n6hg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a713d876ac558d6be4fd1a485e107af4f9b2f3f4/pkg/roachpb/span_config.proto#L182)):*
> ```protobuf
>   // If the host tenant is the source and the target is unspecified then the
>   // associated system span configuration applies over all ranges in the system
>   // (including those belonging to secondary tennats).
> ```

Typo: "tenants"
___
*[pkg/roachpb/span_config.proto, line 205 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-Mvttj4B41VzGi0jQ5V9:-Mvttj4B41VzGi0jQ5VA:b-g3xnjb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a713d876ac558d6be4fd1a485e107af4f9b2f3f4/pkg/roachpb/span_config.proto#L205)):*
> ```protobuf
>   reserved 1;
> 
>   // Target specifies the target (keyspan(s)) the config applies over.
> ```

[nit] Drop "keyspan(s)". Callers can jump to the target definition itself to figure out what the internals represent.
___
*[pkg/roachpb/span_config.proto, line 239 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvtuL4P8-kYW_TSVz8Z:-MvtuL4P8-kYW_TSVz8_:b-dewz2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/a713d876ac558d6be4fd1a485e107af4f9b2f3f4/pkg/roachpb/span_config.proto#L239)):*
> ```protobuf
> 
> // UpdateSpanConfigsRequest is used to update the span configurations and system
> // span configurations over the given spans.
> ```

[nit] Talk about it in terms of targets and span configs, instead of spans. IIUC, we don't have a "system span configuration" here, but we do have this special target concept instead.
___
*[pkg/roachpb/span_config.proto, line 254 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvtuloXDt9k5DwcfG1v:-MvtuloXDt9k5DwcfG1w:bka0gek) ([raw file](https://github.com/cockroachdb/cockroach/blob/a713d876ac558d6be4fd1a485e107af4f9b2f3f4/pkg/roachpb/span_config.proto#L254)):*
> ```protobuf
> // are, however, allowed to overlap across lists. This is necessary to support
> // the delete+upsert semantics described above.
> // Targets of configurations being added must not overlap with any existing
> ```

I'm not sure I follow this last sentence. Is it necessary given your changes above?


<!-- Sent from Reviewable.io -->

arulajmani committed:
spanconfig: remove panic from code around targets

Changed to return an error to the caller when dealing with proto
to target conversions for protos received over the wire. There isn't
likely to be an error in this codepath, but if there is, it makes sense
to return it up the RPC chain to the tenant as opposed to letting it
bring a KV node down.

Release note: None

arulajmani reviewed (commented):
TFTR! 

Will merge on green fr this time. 

> Dismissed @adityamaru and @irfansharif from 11 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-:-MvuZ0rR5jup-1qyuIcx:bahutu1)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @adityamaru, @ajwerner, @irfansharif, and @nvanbenschoten)

___
*[pkg/ccl/kvccl/kvtenantccl/connector.go, line 465 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvtiUQEBax7aamPJH2f:-MvuViE16R7hY3KqzUg3:b-v6s1o3) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/ccl/kvccl/kvtenantccl/connector.go#L465)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Do we need to plumb down a context for a type-conversion
</blockquote></details>

Removed. 
___
*[pkg/roachpb/span_config.proto, line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvMxNf6dmwJguHlaDTO-r1-230:-Mvu5oDE9DzB2SpeZsK8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/36415d73f730b0862b62aaf126df6fca61d08aa1/pkg/roachpb/span_config.proto#L230)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Bumping this isn't necessary, this is safe to revert (makes for a more confusing git commit history if I'm being pedantic).
</blockquote></details>

Done.
___
*[pkg/roachpb/span_config.proto, line 180 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvpA1l5Cadd85_DNeIR:-MvuDS7W77vj54JmcnOa:b-p25ciq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebec94f69bbf7e1b62f7d536606b737e9bebf081/pkg/roachpb/span_config.proto#L180)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

I think the following semantics are sound/easier to get wrong:
- Only the host tenant is allowed to set a target_tenant_id.
  - If the target_tenant_id is unset and the source_tenant_id == host, applies to all tenants.
  - If a target_tenant_id is set, it applies to the that tenant only (could be the host tenant itself)

This is in contrast to making it _necessary_ for secondary tenants to target themselves. In the alternative above, the whole concept of "targeting" a tenant doesn't extend to secondary tenants, which is what we want.
</blockquote></details>

I have a mild preference to what we have here as opposed to what you're suggesting. Mostly because in your suggestion, the `target_tenant_id` being left unset means different things depending on who the source is -- if the source is the host tenant then the target represents the physical cluster, but if the source is a secondary tenant, then the target represents its logical cluster. Expressing these in "absolute" terms feels more intuitive to me.
___
*[pkg/roachpb/span_config.proto, line 175 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvtsCaQ9ingHUR077oL:-Mvu5QwJ2REi6uV9uwQN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a713d876ac558d6be4fd1a485e107af4f9b2f3f4/pkg/roachpb/span_config.proto#L175)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] Spacing around '=' for gogoproto.customname. Ditto below, where the spacing is inconsistent in another way.
</blockquote></details>

Done.
___
*[pkg/roachpb/span_config.proto, line 205 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-Mvttj4B41VzGi0jQ5V9:-Mvu5c0H3Vm4Eyhjac9t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a713d876ac558d6be4fd1a485e107af4f9b2f3f4/pkg/roachpb/span_config.proto#L205)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] Drop "keyspan(s)". Callers can jump to the target definition itself to figure out what the internals represent.
</blockquote></details>

Done.
___
*[pkg/roachpb/span_config.proto, line 239 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvtuL4P8-kYW_TSVz8Z:-Mvu68MT180eHeSP10Zd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a713d876ac558d6be4fd1a485e107af4f9b2f3f4/pkg/roachpb/span_config.proto#L239)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] Talk about it in terms of targets and span configs, instead of spans. IIUC, we don't have a "system span configuration" here, but we do have this special target concept instead.
</blockquote></details>

Done.
___
*[pkg/roachpb/span_config.proto, line 254 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvtuloXDt9k5DwcfG1v:-Mvu6rGA1-EU362isYn5:bkl0vi5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a713d876ac558d6be4fd1a485e107af4f9b2f3f4/pkg/roachpb/span_config.proto#L254)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

I'm not sure I follow this last sentence. Is it necessary given your changes above?
</blockquote></details>

Yeah, I think the previous paragraph captures what I was trying to say here. Removed. 
___
*[pkg/spanconfig/target.go, line 36 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvtjTOzBP4KdCL210yW:-MvuCf8LDjw6Wq0bAKCA:b-xznq3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/spanconfig/target.go#L36)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Getting rid of this error and opting instead for a panic would help simplify code paths (no handling of unhandle-able errors needed) that bottom out at what is affectively a type cast.
</blockquote></details>

For this one in particular, I think it makes sense to return the error up the RPC chain, back to the tenant, instead of bringing down the KV node. For everything else, I cleaned things up. 
___
*[pkg/spanconfig/target.go, line 89 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvpFGzP0k0IggLYtBAp:-MvuTaM4-oUaQZpAsM3k:b-u45z91) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/spanconfig/target.go#L89)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

+1. Would also remove the need to pass in a ctx for what should otherwise be a simple type cast.
</blockquote></details>

Removed the context passing and switched to panic-ing instead. 
___
*[pkg/spanconfig/target.go, line 102 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvtkL6U2wEZ9WfnEoKs:-MvuImMN17s7Q3ZVdXhI:b-igcari) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/spanconfig/target.go#L102)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Would this be simpler if we defined a .Less on the SystemTarget type itself? Ditto below for the equality check, where we delegate to span.Equal cause it's available. It seems generally useful to define these on the base type instead of inlining it in methods on the union type. That would for e.g. let us define sorts on the base types instead of roundtripping through the union type.
</blockquote></details>

Good point, done for both. 
___
*[pkg/spanconfig/target.go, line 155 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvtkvqT98JsPz-73Xaf:-MvuJZDG4V4PfGYGu1DV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/spanconfig/target.go#L155)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] This could be simplified to "return <condition>".
</blockquote></details>

Done.
___
*[pkg/spanconfig/target.go, line 163 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-Mvtly2h0f9luaUipcUw:-MvuTZrZ5Egw3BbmUzA0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/spanconfig/target.go#L163)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Is it fair to say every spanconfig.Target is going to be represented as a roachpb.SpanConfigTarget over the wire? Ditto for []spanconfig.Targets represented as []roachpb.SpanConfigTarget? If so, [nit] how about the following method signatures?

```go
func (t Target) ToProto() roachpb.SpanConfigTarget { .. } // also dropping ctx and panic-ing instead
func TargetsToProtos([]Target) []roachpb.SpanConfigTarget { ...}
func TargetsFromProtos([]roachpb.SpanConfigTarget) []Target { ...}
```
</blockquote></details>

Done.
___
*[pkg/spanconfig/target.go, line 216 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-Mvtl5ml1mKPNMB3g9cb:-MvuKAXf-2QbGTjgBUsV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/spanconfig/target.go#L216)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] s/RecordsToSpanConfigEntries/RecordsToEntries for symmetry with EntriesToRecords below. Also nice to be a little less wordy.
</blockquote></details>

Done.
___
*[pkg/spanconfig/spanconfigreconciler/reconciler.go, line 287 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-MvtnUMMARvhk6hOjo8r:-MvuU-FGC2_-b_OqMI38:bropwsp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/spanconfig/spanconfigreconciler/reconciler.go#L287)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Does this smell to anyone? The server's internal encoding is leaking through to the client if it's having to be careful about not asking for portions of the keyspace. Ideally the server side code just filters these spans out at some level if we only want it to be targeted through explicit SystemSpanConfigTargets. Is this something we're planning to do anything about?
</blockquote></details>

I see a fair bit of this changing once we introduce a method on the `KVAccessor` to get all system span configurations installed by the host tenant. I agree, it isn't great that this leaks here, but I'm not sure silently swallowing this span is better given the targeted nature of our get request. We can see how it looks once we make changes to the reconciler in earnest. 
___
*[pkg/rpc/auth_test.go, line 153 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-Mvtp-Ry7GzW3h-xBgwd:-MvuJlhj1nbrLx7BJH_0:baf33ic) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/rpc/auth_test.go#L153)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Can we avoid "pointer to a span" here as well? Ditto below for makeSystemTarget. For no other reasons than it being sharper to use, for unclear benefits. If we're doing union types, staring at makeSpanConfigTarget, we could instead have two constructors for each type of base type in the union. IIUC, one or the other argument _has_ to be nil, right? Could even repurpose makeSystemTarget and makeSpan for it.
</blockquote></details>

Good call, done. 
___
*[pkg/rpc/auth_test.go, line 653 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76219#-Mvtq1io5p9DQIyJwT42:-Mvu3lq-CE0io2F8KQBy:b-50ku66) ([raw file](https://github.com/cockroachdb/cockroach/blob/2252c531703498f7f1c4a10ba248ec6b221d99d3/pkg/rpc/auth_test.go#L653)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Is this the same test case as the last one? In fact, I'm seeing multiple duplicated sub-tests here. Was this intentional?
</blockquote></details>

They're different, in that we test updates/deletes/gets separately. It's what we had before my change, I just added system span config targets to it. 


<!-- Sent from Reviewable.io -->

arulajmani commented:
bors r+
